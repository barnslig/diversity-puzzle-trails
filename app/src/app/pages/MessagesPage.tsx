import {
  AppBar,
  Toolbar,
  Typography,
  Box,
  Container,
  LinearProgress,
} from "@mui/material";
import { FormattedMessage } from "react-intl";
import * as React from "react";

import MainNav from "../MainNav";
import MessageCard from "../../common/components/MessageCard";
import useMessages from "../../common/hooks/api/useMessages";
import { Message } from "../../common/types/Message";
import HeroMessage from "../../common/components/HeroMessage";

const MessagesPage = () => {
  const { data: messages } = useMessages();

  // Scroll to the bottom after (new) messages are rendered
  React.useEffect(() => {
    const $html = document.documentElement;
    $html.scrollTop = $html.scrollHeight;
  }, [messages]);

  /**
   * Whether the API has returned zero messages
   */
  const hasNoMessages = messages && messages.data.length === 0;

  return (
    <div>
      <AppBar position="fixed">
        <Toolbar>
          <Typography component="h1" sx={{ flexGrow: 1 }} variant="h6">
            <FormattedMessage
              defaultMessage="Nachrichten"
              description="messages page title"
            />
          </Typography>
        </Toolbar>
      </AppBar>
      <MainNav />
      <Box
        component="main"
        paddingBottom={7}
        paddingTop={10}
        sx={{ display: "flex", alignItems: "flex-end", minHeight: "100vh" }}
      >
        {hasNoMessages ? (
          <HeroMessage
            title={
              <FormattedMessage
                defaultMessage="Es gibt noch keine Nachrichten"
                description="messages page no messages yet title"
              />
            }
            description={
              <FormattedMessage
                defaultMessage="Schau doch spÃ¤ter nochmal vorbei."
                description="messages page no messages yet description"
              />
            }
          />
        ) : messages ? (
          <Container maxWidth="sm">
            {messages.data.map((msg: Message) => (
              <MessageCard key={msg.id} message={msg} />
            ))}
          </Container>
        ) : (
          <LinearProgress sx={{ width: "100%" }} />
        )}
      </Box>
    </div>
  );
};

export default MessagesPage;
