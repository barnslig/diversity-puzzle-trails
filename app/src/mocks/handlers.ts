import { rest } from "msw";

import { ClockApiResponse } from "../common/hooks/api/useClock";
import { ClockState } from "../common/types/Clock";

let clockState: ClockState = "running";

export const handlers = [
  rest.get("/games/:gameId/clock", (req, res, ctx) => {
    const clockRes: ClockApiResponse = require("./data/clock.json");
    clockRes.data.attributes.state = clockState;
    return res(ctx.json(clockRes));
  }),
  rest.post("/games/:gameId/clock", (req, res, ctx) => {
    clockState =
      JSON.parse(req.body as string).data.attributes.state || "running";

    const clockRes: ClockApiResponse = require("./data/clock.json");
    clockRes.data.attributes.state = clockState;
    return res(ctx.json(clockRes));
  }),

  rest.get("/games/:gameId/codes/:codeId", (req, res, ctx) => {
    if (req.params.codeId === "already-used") {
      return res(
        ctx.status(403),
        ctx.json(require("./data/code-already-used.json"))
      );
    }

    if (req.params.codeId === "not-found") {
      return res(
        ctx.status(404),
        ctx.json(require("./data/code-not-found.json"))
      );
    }

    return res(ctx.json(require("./data/code.json")));
  }),
  rest.post("/games/:gameId/codes/:codeId", (req, res, ctx) => {
    return res(ctx.json(require("./data/code.json")));
  }),

  rest.get("/games/:gameId/parameters", (req, res, ctx) => {
    return res(ctx.json(require("./data/parameters.json")));
  }),
];
