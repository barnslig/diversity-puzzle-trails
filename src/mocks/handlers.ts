import { rest } from "msw";

export const handlers = [
  rest.get("/games/:gameId/clock", (req, res, ctx) => {
    return res(ctx.json(require("./data/clock.json")));
  }),
  rest.patch("/games/:gameId/clock", (req, res, ctx) => {
    return res(ctx.json(require("./data/clock.json")));
  }),

  rest.get("/games/:gameId/codes/:codeId", (req, res, ctx) => {
    if (req.params.codeId === "already-used") {
      return res(
        ctx.status(403),
        ctx.json(require("./data/code-already-used.json"))
      );
    }

    if (req.params.codeId === "not-found") {
      return res(
        ctx.status(404),
        ctx.json(require("./data/code-not-found.json"))
      );
    }

    return res(ctx.json(require("./data/code.json")));
  }),
  rest.post("/games/:gameId/codes/:codeId", (req, res, ctx) => {
    return res(ctx.json(require("./data/code.json")));
  }),

  rest.get("/games/:gameId/parameters", (req, res, ctx) => {
    return res(ctx.json(require("./data/parameters.json")));
  }),
];
