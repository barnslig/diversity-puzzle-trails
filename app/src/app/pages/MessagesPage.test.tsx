import { rest } from "msw";
import { screen, waitFor } from "@testing-library/react";
import * as React from "react";

import { server } from "../../mocks/server";
import MessagesPage from "./MessagesPage";
import render from "../../common/testing/render";

it("renders a loading screen", () => {
  render(<MessagesPage />);

  expect(screen.getByRole("progressbar")).toBeInTheDocument();
});

it("renders messages", async () => {
  const { container } = render(<MessagesPage />);

  await waitFor(() => expect(screen.getByText(/Test 123/)).toBeInTheDocument());
  expect(container.firstChild).toMatchSnapshot();
});

it("renders no messages yet hero message", async () => {
  server.use(
    rest.get("/games/:gameId/messages", (req, res, ctx) => {
      return res(
        ctx.json({
          data: [],
        })
      );
    })
  );

  const { container } = render(<MessagesPage />);

  await waitFor(() =>
    expect(
      screen.getByText(/Es gibt noch keine Nachrichten/)
    ).toBeInTheDocument()
  );
  expect(container.firstChild).toMatchSnapshot();
});
