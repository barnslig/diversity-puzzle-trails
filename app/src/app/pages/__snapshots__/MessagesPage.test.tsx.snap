// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders messages 1`] = `
<div>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-11b3ww9-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1t29gy6-MuiToolbar-root"
    >
      <h1
        class="MuiTypography-root MuiTypography-h6 css-1juivf6-MuiTypography-root"
      >
        Nachrichten
      </h1>
    </div>
  </header>
  <nav
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 css-avw9ge-MuiPaper-root"
  >
    <div
      class="MuiBottomNavigation-root css-16lloyr-MuiBottomNavigation-root"
    >
      <button
        class="MuiButtonBase-root MuiBottomNavigationAction-root Mui-selected css-1ee5err-MuiButtonBase-root-MuiBottomNavigationAction-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="HomeIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
          />
        </svg>
        <span
          class="MuiBottomNavigationAction-label Mui-selected css-imwso6-MuiBottomNavigationAction-label"
        >
          Status
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiBottomNavigationAction-root css-1ee5err-MuiButtonBase-root-MuiBottomNavigationAction-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M9.5 6.5v3h-3v-3h3M11 5H5v6h6V5zm-1.5 9.5v3h-3v-3h3M11 13H5v6h6v-6zm6.5-6.5v3h-3v-3h3M19 5h-6v6h6V5zm-6 8h1.5v1.5H13V13zm1.5 1.5H16V16h-1.5v-1.5zM16 13h1.5v1.5H16V13zm-3 3h1.5v1.5H13V16zm1.5 1.5H16V19h-1.5v-1.5zM16 16h1.5v1.5H16V16zm1.5-1.5H19V16h-1.5v-1.5zm0 3H19V19h-1.5v-1.5zM22 7h-2V4h-3V2h5v5zm0 15v-5h-2v3h-3v2h5zM2 22h5v-2H4v-3H2v5zM2 2v5h2V4h3V2H2z"
          />
        </svg>
        <span
          class="MuiBottomNavigationAction-label css-imwso6-MuiBottomNavigationAction-label"
        >
          Scanner
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiBottomNavigationAction-root css-1ee5err-MuiButtonBase-root-MuiBottomNavigationAction-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ChatBubbleIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 2H4c-1.1 0-2 .9-2 2v18l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2z"
          />
        </svg>
        <span
          class="MuiBottomNavigationAction-label css-imwso6-MuiBottomNavigationAction-label"
        >
          Nachrichten
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </nav>
  <main
    class="MuiBox-root css-177ukz7"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm css-cuefkz-MuiContainer-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1al38ny-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            Irgendeine Nachricht :)
          </p>
          <span
            class="MuiTypography-root MuiTypography-caption css-2412zy-MuiTypography-root"
          >
            <div />
          </span>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1al38ny-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            Test 123
          </p>
          <span
            class="MuiTypography-root MuiTypography-caption css-2412zy-MuiTypography-root"
          >
            <div />
          </span>
        </div>
      </div>
    </div>
  </main>
</div>
`;

exports[`renders no messages yet hero message 1`] = `
<div>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-11b3ww9-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1t29gy6-MuiToolbar-root"
    >
      <h1
        class="MuiTypography-root MuiTypography-h6 css-1juivf6-MuiTypography-root"
      >
        Nachrichten
      </h1>
    </div>
  </header>
  <nav
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 css-avw9ge-MuiPaper-root"
  >
    <div
      class="MuiBottomNavigation-root css-16lloyr-MuiBottomNavigation-root"
    >
      <button
        class="MuiButtonBase-root MuiBottomNavigationAction-root Mui-selected css-1ee5err-MuiButtonBase-root-MuiBottomNavigationAction-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="HomeIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
          />
        </svg>
        <span
          class="MuiBottomNavigationAction-label Mui-selected css-imwso6-MuiBottomNavigationAction-label"
        >
          Status
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiBottomNavigationAction-root css-1ee5err-MuiButtonBase-root-MuiBottomNavigationAction-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M9.5 6.5v3h-3v-3h3M11 5H5v6h6V5zm-1.5 9.5v3h-3v-3h3M11 13H5v6h6v-6zm6.5-6.5v3h-3v-3h3M19 5h-6v6h6V5zm-6 8h1.5v1.5H13V13zm1.5 1.5H16V16h-1.5v-1.5zM16 13h1.5v1.5H16V13zm-3 3h1.5v1.5H13V16zm1.5 1.5H16V19h-1.5v-1.5zM16 16h1.5v1.5H16V16zm1.5-1.5H19V16h-1.5v-1.5zm0 3H19V19h-1.5v-1.5zM22 7h-2V4h-3V2h5v5zm0 15v-5h-2v3h-3v2h5zM2 22h5v-2H4v-3H2v5zM2 2v5h2V4h3V2H2z"
          />
        </svg>
        <span
          class="MuiBottomNavigationAction-label css-imwso6-MuiBottomNavigationAction-label"
        >
          Scanner
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiBottomNavigationAction-root css-1ee5err-MuiButtonBase-root-MuiBottomNavigationAction-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ChatBubbleIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 2H4c-1.1 0-2 .9-2 2v18l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2z"
          />
        </svg>
        <span
          class="MuiBottomNavigationAction-label css-imwso6-MuiBottomNavigationAction-label"
        >
          Nachrichten
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </nav>
  <main
    class="MuiBox-root css-177ukz7"
  >
    <main
      class="MuiBox-root css-96m5c5"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthSm css-cuefkz-MuiContainer-root"
      >
        <h2
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
        >
          Es gibt noch keine Nachrichten
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
        >
          Schau doch sp√§ter nochmal vorbei.
        </p>
      </div>
    </main>
  </main>
</div>
`;
